### ISSUE TITLE
Eclipse plugin debug is broken

### ISSUE THREAD

**Issue by iloveeclipse** [spotbugs/spotbugs#153]
Findbugs project was used as a valid Eclipse bundle (it had a valid META-INF/MANIFEST.MF) so that one could debug Eclipse plugin and use hot code replace because all required classes were there in the FindBugs bundle. Now someone removed that manifest and so one can't debug Eclipse plugin anymore. OMG.
Also I see that almost all libraries are now inside the hidden local gradle folder and so aren't on the spotbugs bundle classpath. Is there a way to put them somewhere where the bundle could reference it via MANIFEST.MF?

**Comment by iloveeclipse** [comment]
I can understand that Gradle is cool and modern, but what I see now is a complete, absolute, pointless mess.

Not only the manifest from the findbugs project was removed, the manifest of eclipse plugin is pointing now to non existing libraries in the lib folder, which will be copied there by gradle at the very end of the build. 

So debugging a plugin is effectively not possible anymore. WTF?!?

I can't debug Eclipse plugin anymore and so this project is dead for me.

**Comment by jsotuyod** [comment]
@iloveeclipse I understand your frustration, please relax a bit. It's not no one cares about the Eclipse plugin, but mostly that none of us other than you has experience with them.

The issues you mention have been noted, and we will work on getting them straight.

**Comment by iloveeclipse** [comment]
For whatever reason I've expected that the gradle replaces ant without breaking existing functionality, and now I'm even unable to debug things.

What is needed is: 3 projects (spotbugs, eclipsePlugin and eclipsePluginTests) need valid MANIFEST.MF files pointing to existing libraries. This could be achieved for example if the gradle build would copy those jars or create links from lib/ folders to the local versions of those hidden jar files now located somewhere on every user home directory. But since I have no idea how to make it with gradle, and ant build files are now removed, I can't help here, and so my frustration.

**Comment by jsotuyod** [comment]
@iloveeclipse would you mind giving me some minimum input on how to run / debug the eclipse plugin from eclipse? Thanks in advanced

**Comment by iloveeclipse** [comment]
Import spotbugs and eclipsePlugin projects into Eclipse.
Run -> Debug Configurations -> Eclipse Application -> New -> Debug.

This will fail right now because the spotbugs plugin can't find any of spotbugs main classes and dependencies.

This would not fail if spotbugs project would became the manifest back, the spotbugs plugin would fix the dependency to spotbugs (right now it is findbugs) and all the libraries required by the plugin would be really located at the lib directory.

See spotbugs/eclipsePlugin/doc/building_findbugsplugin.txt

**Comment by iloveeclipse** [comment]
To understand how efficiently debug Eclipse plugins one need to know that Eclipse allows to debug *not packaged* plugins, means: "exploded" jar files, directly as they are in the source tree.

This allows to startup huge Eclipse applications including hundreds of plugins from sources, without waiting for a build, and additionally the same trick allows to change the code of the debugged classes on the fly without re-packaging and restarting everything, just in place during debugging.

To achieve that, Eclipse projects must be properly configured: the dependencies should be given not via .classpath file like for plain Java projects but by properly written MANIFEST.MF. The .classpath usually just contains 2 entries for libraries: JRE CONTAINER variable and org.eclipse.pde.core.requiredPlugins.

To start Eclipse from Eclipse, Eclipse PDE tooling also makes sure that the started application uses "exploded" plugins shape and this can only work if the MANIFEST.MF are the only sources of dependencies, because in the "real" life this is the case.

The only difference in the classpaths is that the jar file generated from the actual source code of the plugin at the build time is replaced by the class folder of the plugin project.

**Comment by jsotuyod** [comment]
@iloveeclipse thanks for the response, got it clear.

What we can do is have a task run once to create / update the manifest and have Eclipse use that. By having a task we can make sure that changes to dependencies are automatically applied across all fronts (manifest, project classpath, generated sites / jars), plus it being run only once (initial import, when you change something) just like you currently need to set the properties file pointing to the eclipse install doesn't seem so bad / such a burden, as long as it's properly documented on the README.

If you are ok with this, I'll get to work, I can probably get it done pretty quickly.

**Comment by iloveeclipse** [comment]
BTW I've just found that the generated .classpath is simply wrong. It should not contain the absolute paths to all possible plugins, but should contain very few entries with the most important one: `<classpathentry kind="con" path="org.eclipse.pde.core.requiredPlugins"/>.`

The next issue related to the new shiny gradle world is #160 . There we have no gradle build files (or I don't know where to search for them), and the generated .classpath is unusable, because it does not contain *any dependencies*. Same here, it should contain `<classpathentry kind="con" path="org.eclipse.pde.core.requiredPlugins"/>.`

Best, you check which .classpath files we have before #40 and either restores them or make the gradle thing generate something usable.

**Comment by iloveeclipse** [comment]
@jsotuyod : is there any way to restore the previous ant builds, at least for eclipse plugin and eclipse tests plugin? I'm currently asking myself for what was this gradle change good at all? What it can do better if it can't do trivial things which were working before - let's count: 

1. FindBugs junit tests disabled
2. Eclipse plugin debugging is broken, no way to even start it
3. Eclipse test plugin broken, code is not compilable anymore
4. I haven't really checked what else is broken because every time I check the code I see new problems.

Sorry, this drives me mad. Changes just "because we can".

**Comment by KengoTODA** [comment]
Maybe I don't understand how to use `eclipsePlugin-test` project;
It seems that `eclipsePlugin-test` has no `build.xml` for Ant, even at 3.0.1 release.

* https://github.com/spotbugs/spotbugs/tree/3.0.1/eclipsePlugin-test

And I cannot find related entry in `eclipsePlugin/build.xml`

* https://github.com/spotbugs/spotbugs/blob/3.0.1/eclipsePlugin/build.xml


@iloveeclipse is it possible to build and test by CLI? Or do I need to launch Eclipse GUI to run test?
If we can test by CLI interface such as Ant and Gradle, it's really nice to keep `master` branch stable, because we can ask Travis CI to test.

**Comment by iloveeclipse** [comment]
@KengoTODA : you can't run tests now, due #160. eclipsePlugin-test setup is broken after move to gradle.

Previously we were able to build and run Eclipse tests, I did it from Eclipse. 

We haven't CLI version but this should be also doable, once we will have fixed the test project setup.

**Comment by jsotuyod** [comment]
Sorry for the long delay. I've been working and got pretty far. My current trouble right now is that Eclipse seems to not be finding the Spotbugs classes when running directly from PDE. It does find the Plugin classes and all libs dependencies, just not SpotBugs.

I came across this document: https://wiki.eclipse.org/FAQ_What_causes_my_plug-in_to_build_but_not_to_load_in_a_runtime_workbench%3F

I checked both `MANIFEST.MF` and `build.properties` and they both seem fine. Actually, before I updated them, the `NoClassDefFoundError` was on dom4j classes, but now they are found.

I tried checking the old findbugs code to see how it was configured, but I don't see an explicit reference to `findbugs.jar` on the [MANIFEST](https://github.com/findbugsproject/findbugs/blob/master/eclipsePlugin/META-INF/MANIFEST.MF#L6), and on `build.properties` I do find [this reference](https://github.com/findbugsproject/findbugs/blob/master/eclipsePlugin/build.properties#L1)  but I don't see how it would work, since it's not part of bin / src, and adding the equivalent on spotbugs (and the `build.properties` for spotbugs project) made no difference at all...

If anyone has any idea it would be most welcomed. I may push my changes later on so you can freely look at it (I still have to fix a couple things I broke on the distribution build generation as part of my tests)

**Comment by iloveeclipse** [comment]
The point is: MANIFEST.MF of the plugin needs a new "spotbugs" bundle id entry as *optional* dependency. The finbugs.jar / spotbugs.jar should not used by PDE at all at debug time, and all the spotbugs classes will be found automatically by PDE. If deploying the plugin, we must put the spotbugs classes into the plugin jar, along with the plugin classes (I think this is already done), and so we don't need to put an extra spotbugs.jar on plugin lib/ or any other location.

**Comment by jsotuyod** [comment]
Yes, on the distribution package we merge the classes (with some exclussions), copied from the old ant build.

I can't get it to work on PDE. It is able to load the plugin classes and the dependencies, but not spotbugs itself... The eclipse project has the spotbugs project as a dependency, but still seems unable to locate it's classes. I just submitted a PR so you can have a peek.
